lib/src/revset.rs

```diff
@@ -2293,28 +2295,58 @@ mod tests {
             @r###"Expression("Expected expression of string pattern")"###);
         insta::assert_debug_snapshot!(
             parse(r#"author("foo@")"#).unwrap(),
-            @r###"Filter(Author(Substring("foo@")))"###);
+            @r###"
+        Filter(
+            Author(
+                Substring(
+                    "foo@",
+                    Sensitive,
+                ),
+            ),
+        )
+        "###);
```

Comment:
It might be nice if we can make the filter handle short two‐parameter structures like this; not sure the best way to do it, though.

Reply:
Maybe it's simpler to add `ExactIcase`, `SubstringIcase`, etc. as separate variants?

Most of the users handle `Exact(_, Sensitive)`/`Exact(_, Insensitive)` separately, and it makes some sense that `exact()`/`is_exact()` implies case-sensitive match.

Reply:
Yeah, that makes sense. The other strategies I had toyed around with were putting the case‐sensitivity in a wrapper struct, which didn’t seem better, and representing every pattern as a glob, which I kind of liked for implementation uniformity and maximum expressivity; all the existing operations can still be supported by e.g. checking that there aren’t any unescaped meta‐characters for conversion to an exact string. Of course, if we switched to supporting regexes that would get more complex. Separate constructors might be the best thing for now.

Reply:
I sketched up the separate constructors version locally and it definitely feels nicer than what I have in this PR; I’ll hold off on pushing it until the other stuff is hashed out.

Reply:
> The other strategies I had toyed around with were putting the case‐sensitivity in a wrapper struct, which didn’t seem better, and representing every pattern as a glob,

Beware that `StringPattern` is the type representing a pattern expression, and is passed to revset engine. It's similar to `RevsetExpression`/`FilesetExpression`, and the evaluation backend will do whatever transformation it needs there.

It might be convenient if the pattern had `.compile_glob()`/`.compile_regex()`, but I wouldn't apply heavy preprocessing to `StringPattern` itself.